pixeltable/utils/arrow.py

```diff
@@ -0,0 +1,75 @@
+import pyarrow as pa
+from typing import Any, Dict, Iterator
+import pixeltable.type_system as ts
+
+def get_pixeltable_type(arrow_type : pa.DataType) -> 'pixeltable.types.PixelType':
+    mapping = {
+        pa.string(): ts.StringType(nullable=True),
+        pa.timestamp('us'): ts.TimestampType(nullable=True),
+        pa.bool_(): ts.BoolType(nullable=True),
+        pa.uint8() : ts.IntType(nullable=True),
+        pa.int8() : ts.IntType(nullable=True),
+        pa.uint32() : ts.IntType(nullable=True),
+        pa.uint64() : ts.IntType(nullable=True),
+        pa.int32(): ts.IntType(nullable=True),
+        pa.int64(): ts.IntType(nullable=True),
+        pa.float32(): ts.FloatType(nullable=True),
+    }
+
+    if arrow_type in mapping:
+        return mapping[arrow_type]
+    elif isinstance(arrow_type, pa.FixedShapeTensorType):
+        assert arrow_type.value_type in mapping, f'unknown value type {arrow_type.value_type} for array'
+        return ts.ArrayType(shape=arrow_type.shape, dtype=get_pixeltable_type(arrow_type.value_type))
+    else:
+        assert False, f'default conversion for {arrow_type} not implemented'
+
+def get_arrow_type(pixeltable_type : ts.ColumnType) -> pa.DataType:
+    mapping = {
+        ts.StringType: pa.string(),
+        ts.TimestampType: pa.timestamp('us'), # postgres timestamp is microseconds
+        ts.BoolType: pa.bool_(),
+        ts.IntType: pa.int64(),
+        ts.FloatType: pa.float32(),
+        ts.JsonType: pa.string(), # TODO(orm) pa.struct() is possible
+        ts.ImageType: pa.binary(), # inline image
+        ts.AudioType: pa.string(), # path
+        ts.VideoType: pa.string(), # path
+        ts.DocumentType: pa.string(), # path
+    }
+
+    if pixeltable_type.__class__ in mapping:
+        return mapping[pixeltable_type.__class__]
+    elif isinstance(pixeltable_type, ts.ArrayType):
+        return pa.fixed_shape_tensor(pa.from_numpy_dtype(pixeltable_type.numpy_dtype()), pixeltable_type.shape)
+    else:
+        assert False, f'default conversion for {pixeltable_type} not implemented'
+
+def get_pixeltable_schema(arrow_schema : pa.Schema) -> Dict[str, Any]:
+    return {field.name: get_pixeltable_type(field.type) for field in arrow_schema}
+
+def get_arrow_schema(pixeltable_schema : Dict[str, Any]) -> pa.Schema:
+    return pa.schema({name: get_arrow_type(dtype) for name, dtype in pixeltable_schema.items()})
+
+def to_pydict(batch : pa.RecordBatch) -> Dict[str, Any]:
```

Comment:
so it's Dict[str, List[Any]]?

Reply:
except for arrays, which are kept in one piece, so are not a List[Any]. But I guess are all iterable? so will use that.